keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,432 @@
+  5. Verify that the container keeps executing code while sleeping.
+
+- Container exit/crash testing
+  1. Create a simple pod with a container that will exit soon.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration longer than the container's lifecycle.
+  3. Exit the container and observe the time it takes for the pod to terminate.
+  4. Verify that the pod be deleted successfully without waiting for the entire sleep duration.
+
+- Interaction with termination grace period
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Set the termination grace period to various values, including:
+     - Equal to the sleep duration
+     - Greater than the sleep duration
+     - Greater than the sleep duration, but reduced to less than the sleep duration at runtime
+  4. For each termination grace period value, delete the pod and observe the time it takes for the container to terminate.
+  5. Verify that the container is terminated after the min(sleep, grace).
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit/e2e tests completed and enabled
+- Documentation is added to demonstrate why this is useful in nginx scenario and how exactly nginx needs to be configured.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Additional e2e tests are completed(if needed)
+
+#### GA
+
+- No negative feedback
+- No bug issues reported
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+The previous PreStop behavior will not be broken. Users can continue to use their hooks as it is.
+To use this enhancement, users need to enable the feature gate, and add sleep action in their prestop hook.
+
+#### Downgrade
+Kube-apiserver will ignore sleep in prestop hook.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLifecycleSleepAction
+  - Components depending on the feature gate: kubelet,kube-apiserver
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+No
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+The feature can be disabled in Alpha and Beta versions by restarting kube-apiserver with the feature-gate off. In terms of Stable versions, users can choose to opt-out by not setting the sleep field.
```

Comment:
If users have set the sleep field and then the feature is disabled, it will just be ignored and the old behavior (i.e. no sleep) would apply?

Reply:
> If users have set the sleep field and then the feature is disabled, it will just be ignored and the old behavior (i.e. no sleep) would apply?

Yes, in that case, the prestop hook will not take effect.

Reply:
For most API changes, the feature gate controls the admission of new uses to the system, and not the actual implementation logic.  

In other words, if you enable the feature, use the feature, then disable the feature - it keeps working, bt no NEW uses of the feature work.  This is not universally true, we actually do not have a consistent rule here.